### PR TITLE
API reference for resource scoping

### PR DESCRIPTION
**Pull request by magnatelee**
@bryevdv I haven't found a good way to document properties of dataclasses. Any suggestions?

### CODE REVIEW

**Comment by bryevdv** [comment]
I don't really have much experience with `autosummary`. Does using `.. autoclass::` explicitly for each class yield better results?

File: legate/core/machine.py [hunk 1]
```diff
@@ -75,12 +75,41 @@ def create_empty_range(kind: ProcessorKind) -> ProcessorRange:
 
     @property
     def empty(self) -> bool:
+        """
+        Indiciates if the processor range is empty
```
Comment (manopapad) [inline legate/core/machine.py:79]:
```suggestion
        Indicates if the processor range is empty
```

File: legate/core/machine.py [hunk 2]
```diff
@@ -170,22 +243,66 @@ def __eq__(self, other: object) -> bool:
...
         return self._preferred_kind
 
     @property
     def kinds(self) -> tuple[ProcessorKind, ...]:
+        """
+        Returns the kinds of processors available in this machine
+
+        Returns
+        -------
+        tuple[ProcessorKind]
```
Comment (manopapad) [inline legate/core/machine.py:263]:
```suggestion
        tuple[ProcessorKind, ...]
```

File: legate/core/machine.py [hunk 3]
```diff
@@ -271,6 +452,17 @@ def __and__(self, other: Machine) -> Machine:
 
     @property
     def empty(self) -> bool:
+        """
+        Indiciates if the machine is empty
```
Comment (manopapad) [inline legate/core/machine.py:456]:
```suggestion
        Indicates if the machine is empty
```

File: docs/legate/core/source/api/machine.rst [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+============================
+
+By default, each Legate operation is allowed to use the entire machine for
+parallelization, but oftentimes client programs want control on the machine
+resource assigned to each section of the program. Legate provides a
+programmatic way to control resource assignment, called a *resource scoping*.
+
+To scope the resource, a client takes two steps. First, the client queries the
+machine resource available in the given scope and shrinks it to a subset to
+assign to a scope. Then, the client assigns that subset of machine to a scope
```
Comment (manopapad) [inline docs/legate/core/source/api/machine.rst:15]:
```suggestion
assign to a scope. Then, the client assigns that subset of the machine to a scope
```

File: docs/legate/core/source/api/machine.rst [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  with subset:
+    ...
+
+The machine available to a nested scope is always a subset of that for the
+outer scope. If the machine given to a scope has some resources that are not
+part of the machine for the outer scope, they will be removed during the
+resource scoping. The machine used in a scoping must not be empty; otherwise,
+an ``EmptyMachineError`` will be raised.
+
+In case where a machine of the scope has more than one kind of processors, the
```
Comment (manopapad) [inline docs/legate/core/source/api/machine.rst:36]:
```suggestion
In cases where a machine has more than one kind of processor, the
```

**Review by manopapad — APPROVED** [review]

File: legate/core/machine.py (hunk 1 above)
Comment (magnatelee) [reply legate/core/machine.py]: https://github.com/nv-legate/legate.core/pull/857/commits/7452a8c0ff99b4dcb513c33189e3bcc2f37f3c3f

File: legate/core/machine.py (hunk 2 above)
Comment (magnatelee) [reply legate/core/machine.py]: https://github.com/nv-legate/legate.core/pull/857/commits/7452a8c0ff99b4dcb513c33189e3bcc2f37f3c3f

File: legate/core/machine.py (hunk 3 above)
Comment (magnatelee) [reply legate/core/machine.py]: https://github.com/nv-legate/legate.core/pull/857/commits/7452a8c0ff99b4dcb513c33189e3bcc2f37f3c3f

File: docs/legate/core/source/api/machine.rst (hunk 5 above)
Comment (magnatelee) [reply docs/legate/core/source/api/machine.rst]: https://github.com/nv-legate/legate.core/pull/857/commits/7452a8c0ff99b4dcb513c33189e3bcc2f37f3c3f

File: docs/legate/core/source/api/machine.rst (hunk 4 above)
Comment (magnatelee) [reply docs/legate/core/source/api/machine.rst]: https://github.com/nv-legate/legate.core/pull/857/commits/7452a8c0ff99b4dcb513c33189e3bcc2f37f3c3f

**Comment by magnatelee** [comment]
`autoclass` seems to be able to handle attributes better, but somehow I can't get the detail pages for methods and attributes generated, so I'll probably give it a try in a separate PR.

**Pull request merged**